### DISCUSSION TITLE
Chapter 4 질문 정리

### DISCUSSION THREAD

**Discussion by pullingoff** [JS-OS-study/hyokyung-and-students#11] [General]
4장 질문, 감상을 정리하는 쓰레드입니다.

**Comment by 1chz** [comment]
# 느낀점

---

- `프로세스`가 어떤식으로 생성되는지 대략적이나마 알게 됐다
- HTTP를 통한 프로세스간의 통신이 IPC의 일종이라는 것을 알게 됐다. (소켓 통신이 IPC의 일종이며, HTTP는 소켓 통신의 일종)

# 문제

---

- 비정상 종료되어 결국 종료되지 않고 컴퓨터 자원만 점유하는 프로세스를 ____________라 하며, 이는 컴퓨터를 장기간 켜 놓을 경우 컴퓨터가 느려지는 원인 중 하나가 된다.

**Reply by sihyung92** [reply]
좀비 프로세스!! 👾

**Reply by inseo24** [reply]
zombie!

**Reply by pullingoff** [reply]
좀비 프로세스

**Comment by inseo24** [comment]
<h3>느낀 점 & 궁금한 점 </h3>

- 원래는 각 프로세스마다 영역과 자원을 달리 쓰는 줄 알았는데 새로 프로세스 생성된 초기에는 부모 프로세스의 자원을 공유하기도 하구나
- message passing에는 Direct와 Indirect communication이 있는데 예를 들면 어떤게 있을까?

<hr />
<h3>문제</h3>

- 주소 공간을 공유하기 위한 방법 중 하나로 shared memory는 _______을 통해 ____에 메모리 공간을 요청한다.

**Reply by 1chz** [reply]
주소 공간을 공유하기 위한 방법 중 하나로 shared memory는 `시스템 콜`을 통해 `운영체제`에 메모리 공간을 요청한다.

**Reply by sihyung92** [reply]
시스템콜, 커널(OS)

**Reply by pullingoff** [reply]
`시스템 콜`을 통해 `운영체제 커널`

**Comment by sihyung92** [comment]
# 느낀 점 & 궁금한 점 
- 여러 종류의 자식 프로세스를 거느리는 부모 프로세스엔 어떤 것이 있을까?  -> 젠킨스가 생성하는 Job이 프로세스를 생성하는 방식으로 동작한다.
- 포트를 통한 데이터 전달은 자주 보았는데, 직접 프로세스로 전달하는 Direct Commnication이 궁금하다

# 문제
- 유닉스 OS 에서 프로세스 생성과 관련된 시스템 콜은 자식 프로세스를 생성하는 ____, 새로운 프로세스를 덮어쓰는 ____, 자식 프로세스가 끝날때 까지 대기하는 ____, 모든 자원을 할당 해제하는 ____가 있다.

**Reply by 1chz** [reply]
유닉스 OS 에서 프로세스 생성과 관련된 시스템 콜은 자식 프로세스를 생성하는 `fork()`, 새로운 프로세스를 덮어쓰는 `exec()`, 자식 프로세스가 끝날때 까지 대기하는 `wait()`, 모든 자원을 할당 해제하는 `exit()`가 있다.

**Reply by inseo24** [reply]
fork(), exec(), wait(), exit()

**Reply by pullingoff** [reply]
fork, exec, wait, exit

**Comment by pullingoff** [comment]
# 느낀 점
- 부모 프로세스가 자식 프로세스를 만드는 건 알았는데 구체적인 방법과 그때 사용되는 명령어를 알게 돼서 지식이 깊어진 느낌이다.
- 지난 챕터들에서 배웠던 프로그램 카운터나 blocked 등의 개념이 이번 장에도 연결돼 나와서 엮어가며 생각해보는 맛이 있었다.

# 문제
1. message passing 방식에선 공유 변수 일체 없이 _____을 통해 메시지를 보낸다.
2. 프로세스를 관리할 때 wait 명령어는 주로 ___ 프로세스가 ___ 프로세스가 종료되길 기다린다.

**Reply by 1chz** [reply]
message passing 방식에선 공유 변수 일체 없이 `커널`을 통해 메시지를 보낸다.

프로세스를 관리할 때 wait 명령어는 주로 `부모` 프로세스가 `자식` 프로세스가 종료되길 기다린다.

**Reply by inseo24** [reply]
1. 커널
2. 부모, 자식

**Reply by sihyung92** [reply]
1. 커널
2. 부모, 자식
